src/Qir/Runtime/stdlib/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#![deny(clippy::all, clippy::pedantic)]
+
+// Rust Implementation for Quantum Intermediate Representation
+
+pub mod arrays;
+pub mod bigints;
+pub mod callables;
+pub mod math;
+pub mod output_recording;
+pub mod range_support;
+pub mod strings;
+pub mod tuples;
+
+use std::{
+    ffi::CString,
+    rc::{Rc, Weak},
+};
+
+/// Utility used for managing refcounted items.
+unsafe fn update_counts<T>(raw_rc: *const T, update: i32, is_alias: bool) {
+    let mut remaining = update;
+    while remaining != 0 {
+        let rc = Rc::from_raw(raw_rc);
+        if remaining > 0 {
+            // Create and leak new instances to increment the count on the contained item.
+            if is_alias {
+                let _ = Weak::into_raw(Rc::downgrade(&rc));
+            } else {
+                let _ = Rc::into_raw(Rc::clone(&rc));
```

Comment:
Are there reasons to considuer using `Arc` for atomics support; does it have weak references?

Reply:
After some discussion we landed at leaving this as an `Rc` for now and not speculatively add some parts of multithreaded support.